Product Collection: Add loading indicator for client-side pagination

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR introduces an animation(loading indicator) for client-side pagination in the Product Collection block. Inspired by the animation used in the Query Loop block, this feature aims to enhance user experience by displaying an animation if the navigation to a new page exceeds 400ms. For more information, refer to issue https://github.com/woocommerce/woocommerce/issues/44393.

Closes #44393

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add a Product Collection block to a new post & choose any collection. 
2. Adjust the items per page in the toolbar controls to "3" to ensure only three products are displayed per page, creating multiple pages for navigation.
3. Publish the post & verify on Frontend that:
    1. An animation appears when changing pages through pagination if the new page's load time surpasses approximately 400ms, signaling that the next page is loading.
    2. Tip: Consider throttling the network speed to "Slow 3G" to simulate slower loading times.

**Sidenote:** It's recommended to test the frontend in incognito mode. This is because certain plugins might add a top navbar, potentially obscuring the animation. Testing in incognito mode helps ensure that the animation remains visible.

Here is a quick demo:

https://github.com/woocommerce/woocommerce/assets/16707866/c18d1700-e6c3-422e-bb1f-cb468382bead



<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Product Collection: Add loading indicator for client-side pagination

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

imanish003 committed:
Add animation for client-side pagination

This includes:
- Addition of animation state management in the frontend file to control the visual transition between pagination states.
- Introduction of new SCSS rules for the start and finish animations, ensuring a seamless and visually appealing pagination experience.
- Modification of the PHP logic to inject necessary HTML for the animation to be applied.

These updates aim to provide a more engaging and responsive interface for users navigating through product collection.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/44393-product-collection-implement-animation-for-page-changes

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/style.scss:
```diff
@@ -0,0 +1,48 @@
+@keyframes wc-block-product-collection__pagination-start-animation {
+	0% {
+		transform: scaleX(0);
+		transform-origin: 0% 0%;
```
Nitpick - `transform-origin: 0% 0%;` can be moved to `.wc-block-product-collection__pagination-animation` since it doesn't change during animation and it's repeated in `100%` definition

imanish003 replied:
Good catch! I have made the changes in https://github.com/woocommerce/woocommerce/pull/44571/commits/22603a8673517036b1de7b744b54a425fc68fab1

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/style.scss:
```diff
@@ -0,0 +1,48 @@
+	max-width: 100vw;
+	height: 4px;
+	background-color: var(--wp--preset--color--primary, #000);
+	opacity: 0;
```
What do you think about adding `pointer-events: none` to this element? It's always there in a DOM and with some minimalistic design it may be covering some interactive elements (like in the screenshot below, where I intentionally made it green). I know it's super edge case for someone to click in the top 4px but doesn't hurt to add it as `pointer-events: none` allows passing clicks to the interactive elements "below".

<img width="847" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/df4b7aac-9d98-44af-bf70-2aea674e2d1f">

imanish003 replied:
Good catch! I have made the changes in https://github.com/woocommerce/woocommerce/pull/44571/commits/22603a8673517036b1de7b744b54a425fc68fab1 🙌🏻

Looks like this could be improved in Query Loop animation too 🙂

kmanijak reviewed (approved):
I left some minor comments but pre-approving!

One difference I spotted between Query Loop block and Product Collection is in Query Loop animation appears only on mobile. Do you know the reason they omitted desktop? Asking just in case there was important reason which we may need to take into account as well 🤔

imanish003 commented:
> Do you know the reason they omitted desktop?

For me, it's working on Desktop too. Can you please try in incognito mode as I mentioned in this PR description too? 

> Sidenote: It's recommended to test the frontend in incognito mode. This is because certain plugins might add a top navbar, potentially obscuring the animation. Testing in incognito mode helps ensure that the animation remains visible.

kmanijak commented:
> For me, it's working on Desktop too.

You're right, never mind!

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/44393-product-collection-implement-animation-for-page-changes

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

imanish003 committed:
Allow user clicks under product collection's loading animation

This commit enhances the user experience of the loading animation for the product collection block. Changes include:

- Specifying `transform-origin: 0% 0%;` directly within the block's initial style to indicate the animation should start from the left
- Adding `pointer-events: none;` to allow user interactions with elements underneath the loading animation, thus improving usability by not blocking clicks.

Additionally, redundant `transform-origin` properties were removed from the `@keyframes` declaration to clean up the code and avoid unnecessary repetition. This simplification contributes to both the maintainability and readability of the stylesheet.

imanish003 committed:
Merge branch 'add/44393-product-collection-implement-animation-for-page-changes' of https://github.com/woocommerce/woocommerce into add/44393-product-collection-implement-animation-for-page-changes

imanish003 committed:
Fix linting errors in SCSS file
